src/Transpile.ts

```diff
@@ -0,0 +1,192 @@
+import * as ts from "typescript";
+import { CompilerOptions } from "./CompilerOptions";
+import { LuaPrinter } from "./LuaPrinter";
+import { LuaTransformer } from "./LuaTransformer";
+import { TranspileError } from "./TranspileError";
+
+function getCustomTransformers(
+    options: CompilerOptions,
+    customTransformers: ts.CustomTransformers,
+    onSourceFile: (sourceFile: ts.SourceFile) => void
+): ts.CustomTransformers {
+    // TODO: https://github.com/Microsoft/TypeScript/issues/28310
+    const forEachSourceFile = (
+        node: ts.SourceFile,
+        callback: (sourceFile: ts.SourceFile) => ts.SourceFile
+    ) =>
+        ts.isBundle(node)
+            ? ((ts.updateBundle(node, node.sourceFiles.map(callback)) as unknown) as ts.SourceFile)
+            : callback(node);
+
+    const luaTransformer: ts.TransformerFactory<ts.SourceFile> = () => node =>
+        forEachSourceFile(node, sourceFile => {
+            onSourceFile(sourceFile);
+            return ts.createSourceFile(sourceFile.fileName, "", ts.ScriptTarget.ESNext);
+        });
+
+    return {
+        afterDeclarations: customTransformers.afterDeclarations,
+        before: [
+            ...(customTransformers.before || []),
+            ...(customTransformers.after || []),
+            luaTransformer,
+        ],
+    };
+}
+
+export interface TranspiledFile {
+    lua?: string;
+    sourceMap?: string;
+    declaration?: string;
+    declarationMap?: string;
+}
+
+export interface TranspilationResult {
```

Comment:
Thoughts on making this something like the following?
```ts
interface FailTranspilationResult { success: false, diagnostics: ts.Diagnostic[] }
interface SuccessTranspilationResult { success: true, diagnostics: ts.Diagnostic[], transpiledFiles: Map<string, TranspiledFile> }
type TranspilationResult = FailTranspilationResult | SuccessTranspilationResult;
```

Reply:
IMO such results aren't really easy to use, because they require checking for a status, even if you don't really care about it. Also TS doesn't discriminate such unions with standard conditions and requires quite ugly comparisons with boolean literals.
If you think there's a need to return a status, I'd rather add `emitSkipped: boolean;` like in `program.emit`, but I don't really see much use in it, especially considering that we don't use a callback-based emit and all the required state to do any checks is available.